src/gadgets/tower_extension/fq12.rs

```diff
@@ -0,0 +1,686 @@
+use std::sync::Arc;
+
+use pairing::{bn256::Fq as BN256Fq, ff::PrimeField, BitIterator};
+
+use super::{
+    fq2::Fq2,
+    fq6::Fq6,
+    params::{
+        bn256::{BN256Extension12Params, BN256Extension6Params},
+        Extension12Params, Extension6Params,
+    },
+};
+
+use crate::cs::Variable;
+use crate::gadgets::traits::allocatable::CSPlaceholder;
+use crate::gadgets::traits::encodable::CircuitVarLengthEncodable;
+use crate::{
+    cs::traits::cs::ConstraintSystem,
+    field::SmallField,
+    gadgets::{
+        boolean::Boolean,
+        non_native_field::traits::NonNativeField,
+        traits::{
+            allocatable::CSAllocatable, selectable::Selectable, witnessable::WitnessHookable,
+        },
+    },
+};
+
+/// `Fq12` field extension implementation in the constraint system. It is implemented
+/// as `Fq6[w]/(w^2-v)` where `w^6=9+u`. In other words, it is a set of
+/// linear polynomials in a form `c0+c1*w`, where `c0` and `c1` are elements of `Fq6`.
+/// See https://hackmd.io/@jpw/bn254#Field-extension-towers for reference. For
+/// implementation reference, see https://eprint.iacr.org/2006/471.pdf.
+#[derive(Clone, Debug, Copy)]
+pub struct Fq12<F, T, NN, P>
+where
+    F: SmallField,
+    T: PrimeField,
+    NN: NonNativeField<F, T>,
+    P: Extension12Params<T>,
+{
+    pub c0: Fq6<F, T, NN, P::Ex6>,
+    pub c1: Fq6<F, T, NN, P::Ex6>,
+    _marker: std::marker::PhantomData<(F, T)>,
+}
+
+impl<F, T, NN, P> Fq12<F, T, NN, P>
+where
+    F: SmallField,
+    T: PrimeField,
+    NN: NonNativeField<F, T>,
+    P: Extension12Params<T>,
+{
+    /// Creates a new `Fq12` element from two `Fq6` components.
+    pub fn new(c0: Fq6<F, T, NN, P::Ex6>, c1: Fq6<F, T, NN, P::Ex6>) -> Self {
+        Self {
+            c0,
+            c1,
+            _marker: std::marker::PhantomData::<(F, T)>,
+        }
+    }
+
+    pub fn from_c0c3c4<CS>(
+        cs: &mut CS,
+        c0: Fq2<F, T, NN, <<P as Extension12Params<T>>::Ex6 as Extension6Params<T>>::Ex2>,
+        c3: Fq2<F, T, NN, <<P as Extension12Params<T>>::Ex6 as Extension6Params<T>>::Ex2>,
+        c4: Fq2<F, T, NN, <<P as Extension12Params<T>>::Ex6 as Extension6Params<T>>::Ex2>,
+    ) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let zero = Fq2::zero(cs, c0.c0.get_params());
+        let c0 = Fq6::new(c0.clone(), zero.clone(), zero.clone());
+        let c1 = Fq6::new(c3.clone(), c4.clone(), zero);
+
+        Self::new(c0, c1)
+    }
+
+    pub fn pow_u32<CS, S: AsRef<[u64]>>(&mut self, cs: &mut CS, exponent: S) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let mut result = Self::one(cs, self.c0.c0.get_params());
+        let mut found_one = false;
+
+        for i in BitIterator::new(exponent) {
+            if found_one {
+                result = result.square(cs);
+            } else {
+                found_one = i;
+            }
+
+            if i {
+                result = result.mul(cs, self);
+            }
```

Comment:
same thing here, seems like this will cause divergent circuits based on input

Reply:
We are currently using this function for constant exponents, which are the same for all circuits, but I agree that it is unsafe to leave the function further as it is since homogeneity is not guaranteed at this level. I made it homogeneous ;)